python/src/scipp/compat/mantid.py

```diff
@@ -195,15 +195,41 @@ def validate_and_get_unit(unit):
         return known_units[unit]
 
 
-def init_pos(ws):
-    nHist = ws.getNumberHistograms()
-    pos = np.zeros([nHist, 3])
+def _to_spherical(x, y, z):
+    r = np.sqrt((x * x) + (y * y) + (z * z))
+    theta = np.arccos(z / r)
+    phi = np.arctan2(y, x)
+    return r, theta, phi
+
+
+def _to_cartesian(r, theta, phi):
+    x = r * np.sin(theta) * np.cos(phi)
+    y = r * np.sin(theta) * np.sin(phi)
+    z = r * np.cos(theta)
+    return x, y, z
 
+
+def init_pos(ws, sample_pos):
+    import numpy as np
     spec_info = ws.spectrumInfo()
-    for i in range(nHist):
+    det_info = ws.detectorInfo()
+    pos = np.zeros([len(spec_info), 3])
+    for i in range(len(spec_info)):
         if spec_info.hasDetectors(i):
-            p = spec_info.position(i)
-            pos[i, :] = [p.X(), p.Y(), p.Z()]
+            definition = spec_info.getSpectrumDefinition(i)
+            theta, phi, r = 0.0, 0.0, 0.0
+            n_dets = len(definition)
+            for j in range(n_dets):
```

Comment:
Isn't it really slow to do all the math in a loop? Can you just store the input values in an array?

Reply:
`spec_info` is enumerable, so can make a small change around that, but cannot extract ALL positions from spectrum_info without loop. SpectrumDefinition likewise cannot get positions without going via DetectorInfo.

Reply:
Sure, you need the loop for the init, but all the more expensive math could be done on numpy arrays?

Reply:
I pretty much had a working implementation for this, but it got really messy for the following reasons:

1. The spectra-detector mappings have to be dealt with. This means carrying around detector indexes and spectrum indexes and performing a grouping operation for the averaging at some point. 
1. Some spectra have no detectors. The processing route here is then quite different (we only know spectrum numbers, no grouping) so have to avoid grouping etc.
1. It is normally the case that the detector-spectra mapping is fixed, but this is not guaranteed. This affects the ability to preallocate arrays of the right side for the operations. At the very least a shrink is required. This is also true of the problems introduced by having no detectors.

Reply:
Can you use `scipp.groupby` to do this in a nice way?